File: Apps/BankBusiness/backend/service/ContactPersonService.cs
using backend.dto.ContactPerson;
using backend.kafka;
using backend.kafka.requests.ContactPerson;
using backend.model;
using backend.repository;

namespace backend.service;

public class ContactPersonService(
    ContactPersonRepository contactPersonRepository,
    BusinessClientService businessClientService,
    KafkaProducer kafkaProducer)
{
    public List<ContactPersonReadDto> GetContactPersons(string businessClientEmail)
    {
        return contactPersonRepository
            .FindAllByBusinessClientId(businessClientService.GetIdByEmail(businessClientEmail)).Result
            .Select(x => new ContactPersonReadDto(x.Surname, x.Name, x.Patronymic, x.PhoneNumber, x.Email))
            .ToList();
    }

    public ContactPersonReadDto? GetContactPersonById(long personId)
    {
        var person = contactPersonRepository.FindById(personId).Result;
        return person == null
            ? null
            : new ContactPersonReadDto(
                person.Surname,
                person.Name,
                person.Patronymic,
                person.PhoneNumber,
                person.Email
            );
    }

    public async Task CreateContactPerson(string businessClientEmail, ContactPersonCreateDto dto)
    {
        await kafkaProducer.produce("contact-person_create_topic",
            new ContactPersonCreateRequest(
                businessClientEmail,
                dto.Surname,
                dto.Name,
                dto.Patronymic,
                dto.PhoneNumber,
                dto.Email
            )
        );
    }

    public async Task UpdateContactPerson(string businessClientEmail, ContactPersonUpdateDto dto)
    {
        await kafkaProducer.produce("contact-person_update_topic", 
            new ContactPersonUpdateRequest(
                businessClientEmail,
                dto.ContactPersonId,
                dto.Surname,
                dto.Name,
                dto.Patronymic,
                dto.PhoneNumber,
                dto.Email
                )
            );
    }

    public async Task DeleteContactPerson(string businessClientEmail, long personId)
    {
        await kafkaProducer.produce("contact-person_delete_topic", new ContactPersonDeleteRequest(businessClientEmail, personId));
    }

    public bool IsContactPersonLinkedToBusinessClient(string businessClientEmail, long personId)
    {
        return contactPersonRepository.GetBusinessClientEmailByContactPersonId(personId).Result == businessClientEmail;
    }
}

File: Apps/BankBusiness/backend/src/main/scala/service/ContactPersonService.scala
package service

import dao.repository.{BusinessClientRepository, ContactPersonRepository, ContactPersonRepositoryLive}
import dto.contactperson.{ContactPersonCreateDto, ContactPersonReadDto, ContactPersonUpdateDto}
import errors.NoEntityPresented
import kafka.ProducerService
import kafka.messages.contactperson.{ContactPersonCreateRequest, ContactPersonDeleteRequest, ContactPersonUpdateRequest}
import model.ContactPerson
import zio.{RIO, RLayer, Task, ZIO, ZLayer}

case class ContactPersonService(
                                 private val contactPersonRepository: ContactPersonRepository,
                                 private val businessClientService: BusinessClientService) {

  def getContactPersons(businessClientEmail: String): RIO[BusinessClientRepository, List[ContactPersonReadDto]] = for {
    id <- businessClientService.getIdByEmail(businessClientEmail)
    contactPersons <- contactPersonRepository.findAllByBusinessClientId(id)
  } yield contactPersons

  def getContactPersonById(personId: Long): Task[Option[ContactPerson]] =
    contactPersonRepository.findById(personId)

  def getContactPersonDtoByEmail(businessClientEmail: String, personId: Long): RIO[BusinessClientRepository, ContactPersonReadDto] =
    (for {
      id <- businessClientService.getIdByEmail(businessClientEmail)
      dto <- contactPersonRepository.findDtoById(id, personId)
    } yield (dto))
      .someOrFail(NoEntityPresented())

  def createContactPerson(businessClientEmail: String, dto: ContactPersonCreateDto): RIO[ProducerService, Boolean] = ZIO.scoped {
    for {
      kafkaRequest <- ZIO.succeed(ContactPersonCreateRequest(
        businessClientEmail,
        dto.surname,
        dto.name,
        dto.patronymic,
        dto.phoneNumber,
        dto.email
      ))
      metadata <- ZIO.serviceWithZIO[ProducerService](_.produce(kafkaRequest))
      offset <- metadata.map(_.offset()).option
    } yield offset.isDefined
  }

  def updateContactPerson(businessClientEmail: String, dto: ContactPersonUpdateDto): RIO[ProducerService, Boolean] = ZIO.scoped {
    for {
      kafkaRequest <- ZIO.succeed(ContactPersonUpdateRequest(
        businessClientEmail,
        dto.id,
        dto.surname,
        dto.name,
        dto.patronymic,
        dto.phoneNumber,
        dto.email
      ))
      metadata <- ZIO.serviceWithZIO[ProducerService](_.produce(kafkaRequest))
      offset <- metadata.map(_.offset()).option
    } yield offset.isDefined
  }

  def deleteContactPerson(businessClientEmail: String, personId: Long): RIO[ProducerService, Boolean] = ZIO.scoped {
    for {
      kafkaRequest <- ZIO.succeed(ContactPersonDeleteRequest(
        businessClientEmail, personId
      ))
      metadata <- ZIO.serviceWithZIO[ProducerService](_.produce(kafkaRequest))
      offset <- metadata.map(_.offset()).option
    } yield offset.isDefined
  }

  def isContactPersonLinkedToBusinessClient(businessClientEmail: String, personId: Long): RIO[BusinessClientRepository, Boolean] = for {
    id <- businessClientService.getIdByEmail(businessClientEmail)
    res <- contactPersonRepository.isPersonLinkedToClient(id, personId)
  } yield res

}

object ContactPersonService {
  val layer: RLayer[ContactPersonRepository & BusinessClientService, ContactPersonService] =
    ZLayer.fromFunction(ContactPersonService.apply(_, _))
}
